src/service/loading-indicator.js

```diff
@@ -0,0 +1,161 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as dom from '../dom';
+import {createLoaderElement} from './loader-element';
+import {htmlFor} from '../static-template';
+import {registerServiceBuilderForDoc} from '../service';
+
+const MIN_SIZE = 20;
+
+/**
+ * @typedef {{
+ *   shown: boolean,
+ *   loader: !Element,
+ *   container: !Element,
+ * }}
+ */
+let LoadingIndicatorStateDef;
+
+/**
+ * @param {!Node|!./ampdoc-impl.AmpDoc} nodeOrDoc
+ */
+export function installLoadingIndicatorForDoc(nodeOrDoc) {
+  registerServiceBuilderForDoc(
+    nodeOrDoc,
+    'loadingIndicator',
+    LoadingIndicatorImpl
+  );
+}
+
+/**
+ * @implements {../service.Disposable}
+ */
+export class LoadingIndicatorImpl {
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    const inViewport = this.inViewport_.bind(this);
+    /** @private @const {!IntersectionObserver} */
+    this.io_ = new ampdoc.win.IntersectionObserver((records) =>
+      /** @type {!Array<!IntersectionObserverEntry>} */ (records).forEach(
+        inViewport
+      )
+    );
+
+    /** @private @const {!WeakMap<!AmpElement, !LoadingIndicatorStateDef>} */
+    this.states_ = new WeakMap();
+  }
+
+  /** @override */
+  dispose() {
+    this.io_.disconnect();
+  }
+
+  /**
+   * @param {!AmpElement} element
+   */
+  track(element) {
+    this.io_.observe(element);
+  }
+
+  /**
+   * @param {!AmpElement} element
+   */
+  untrack(element) {
+    this.io_.unobserve(element);
+    this.cleanup_(element);
+  }
+
+  /**
+   * @param {!IntersectionObserverEntry} record
+   * @private
+   */
+  inViewport_(record) {
+    const {target, isIntersecting, boundingClientRect} = record;
+    const {width, height} = boundingClientRect;
+    const element = /** @type {!AmpElement} */ (target);
+
+    const show = isIntersecting && width > MIN_SIZE && height > MIN_SIZE;
+
+    let state = this.states_.get(element);
+    const isCurrentlyShown = (state && state.shown) || false;
+    if (show === isCurrentlyShown) {
+      // Loading state is the same.
+      return;
+    }
+
+    if (show && !state) {
+      state = this.createState_(element, width, height);
+      this.states_.set(element, state);
+    }
+    if (state) {
+      state.shown = show;
+      state.container.classList.toggle('amp-hidden', !show);
+      state.loader.classList.toggle('amp-active', show);
+    }
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @param {number} width
+   * @param {number} height
+   * @return {!LoadingIndicatorStateDef}
+   * @private
+   */
+  createState_(element, width, height) {
+    const startTime = Date.now();
```

Comment:
I don't think so. InOb is pretty good at firing when the page is not super busy.

Reply:
In that case we can remove the `startTime` concept - I don't think it serves any purpose if we no longer have a delay

Reply:
Yeah. I just wanted to avoid refactoring spilling deeper into the insides of the actual implementation of the loader. Are you ok I leave it up to some future rewrite?